src/aiida/cmdline/commands/cmd_code.py

```diff
@@ -260,9 +260,14 @@ def export(code, output_file, sort):
         if value is not None:
             code_data[key] = str(value)
 
+    if output_file is None:
+        output_file = f"{code_data['label']}.yml"
```

Comment:
The option defines `Path(exists=False)` and so will error if the file exists. You probably want to apply the same check here. Even better though (since you will be reimplementing the logic of click) we try to make the default actual part of the option definition. Now since it depends on the code label, it cannot be a static default, but there is the concept of "contextual defaults" in click. See here for an example: https://github.com/aiidateam/aiida-core/blob/6db2f4060d4ece4552f5fe757c0f7d938810f4d1/src/aiida/cmdline/commands/cmd_user.py#L57

Reply:
Not sure `contextual_default` really is the best way here? It doesn't seem to be something that's implemented in `click` itself, but in the constructor of the `InteractiveOption` class:
https://github.com/aiidateam/aiida-core/blob/6db2f4060d4ece4552f5fe757c0f7d938810f4d1/src/aiida/cmdline/params/options/interactive.py#L50-L59

And I also only ever saw it used for these interactive options. Instead, I propose we make `OUTPUT_FILE` a globally available `option` under `cmdline/params/options/main.py`, and still set the default similarly as it is done in the implementation of my PR? In addition, I'd actually set `Path(exists=True)`, as I don't think it's a problem to silently overwrite these config YAML files (I'd change the label to the full `code@computer.yml` though).

Furthermore, `--output-file` is also used in `verdi node graph generate`, so I'd replace that there with the global `option` to not have three different ways of achieving the same thing, as `OUTPUT_FILE` also exists as a required, positional `argument` in `cmdline/params/arguments/main.py`. I'm not too happy with having it defined in two places eventually, but I'd still also keep it there, as that is used for `verdi archive create`, for which an automatic naming scheme is not as straightforward as for a `Computer` and `Code` (but happy to learn about a better alternative).

Pinging @mbercx as I mentioned this to him today.

Reply:
TLDR: I wrote a bunch of separate comments, and then realized that really the first question to ask is to make sure you understand the difference between option and argument and whether you are proposing to change the output file _arguments_ for these export commands into output file _options_? If so, I don't think I agree and they should stay arguments. 

-----


You are right, the `contextual_default` is something that we implement, but you could use [`callbacks`](https://click.palletsprojects.com/en/8.1.x/options/#callbacks-for-validation) for the same purpose, which is a `click` native feature.

However, there is a bigger problem. These are both possibilities for _options_ but the current parameters you are changing are _arguments_, i.e., positional parameters, and they do not have the concept of callbacks.

> I propose we make OUTPUT_FILE a globally available option under cmdline/params/options/main.py

Since it is not an option but an argument, if we globalize it, it should go in `aiida.cmdline.params.arguments.main` instead, and there already is the `OUTPUT_FILE`. Or are you suggesting to change it from an argument into an option?

> In addition, I'd actually set Path(exists=True), as I don't think it's a problem to silently overwrite these config YAML files (I'd change the label to the full code@computer.yml though).

Here I don't agree. It could very well be that a user has a _different_ yaml in their CWD that is for a different code/computer but just happens to share the same name as the default that would be generated automatically. I think silently overriding here is bad, _especially_ if specifying the output name is optional.

> Furthermore, --output-file is also used in verdi node graph generate, so I'd replace that there with the global option to not have three different ways of achieving the same thing as OUTPUT_FILE also exists as a required, positional argument in cmdline/params/arguments/main.py. I'm not too happy with having it defined in two places eventually,

Note that here it _has_ to be an option. It actually used to be an argument, but we were forced to change it in order to allow multiple root nodes to be specified through arguments.
This demonstrates that an option and argument are not interchangeable solutions. Sometimes you need on or the other, or one or the other would be preferred.

For commands where the only argument would be an output file, I think it _should_ be an argument and usually it is required. But if we provide a default it technically becomes optional but I think it can be justified to keep it as an argument instead of changing it into an option.

Reply:
The way I understand is that `option`s are specified via a single / double hyphen specifiers, while `argument`s are positional.

I'm not proposing we change them everywhere, and I don't really have an opinion / preference on what to make them. The problem, as you said, is that for arguments the `contextual_default` and `callback` (though, this I didn't know) features don't work, and that was the reason I considered turning it into an `option` instead. If setting the default can be achieved differently (e.g. via the logic I proposed initially), I'm fine with keeping them as (now optional) `argument`s.

@mbercx mentioned he'd like to have the same behavior as for `verdi node graph generate`, though, as the commands are different, I think it's fine that their API is also slightly different.

> Note that here it has to be an option. It actually used to be an argument, but we were forced to change it in order to allow multiple root nodes to be specified through arguments.

This I didn't know, thanks for providing the background (though, scope-creeping here, why not allow for multiple codes to be exported? :p)

Also fine for me to keep failing if the file already exists, so that the user is on the safe side 👍🏽

Reply:
Yeah, I agree it's not very useful without that (and in general) :D